### PR TITLE
New representation for identifiers

### LINKED ISSUE #14: New representation of identifiers: One String (one allocation) plus indices
I think a representation that has just one allocation would be more useful for the vast majority of real-world use cases. We could still store the port as a `u16` separately, but storing the host as `url::Host` seems to be of little benefit and having zero-cost `&str` accessors for parts of, or the whole, identifiers, seems much more important.

### PR DESCRIPTION
**Pull request by jplatte**
Will resolve #14 

Lots of tests currently fail, but I'm unsure whether they were ever a good idea (okay, at least the first one I looked at).

### CODE REVIEW

**Comment by jplatte** [comment]
Note to self:

* implement `TryFrom<String>`, potentially `TryFrom<Cow<str>>`.
* change `u8` to `NonZeroU8`

**Comment by jplatte** [comment]
Unresolved question (for after this PR probably): How hard is it to check whether creating a `url::Host` would succeed, without actually creating it? If we could validate the hostname part of an identifier to that extent, we could provide a non-`Option`al `Host` accessor for most identifier types, *maybe* even storing the `Host` on first access via `OnceCell<Host>`. All of this would have to be behind a feature flag though, the dependencies should not be mandatory if that's all they're needed for.

File: src/room_id_or_room_alias_id.rs [hunk 1]
```diff
@@ -32,92 +26,70 @@ use crate::{
...
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        match *self {
-            RoomIdOrAliasId::RoomAliasId(ref room_alias_id) => {
-                serializer.serialize_str(&room_alias_id.to_string())
-            }
-            RoomIdOrAliasId::RoomId(ref room_id) => serializer.serialize_str(&room_id.to_string()),
+    /// Whether this is a room alias id (stars with `'#'`)
```
Comment (auscompgeek) [inline src/room_id_or_room_alias_id.rs:54]:
```suggestion
    /// Whether this is a room alias id (starts with `'#'`)
```

File: src/room_id.rs [hunk 2]
```diff
@@ -43,77 +34,45 @@ impl RoomId {
...
     }
 
-    /// Returns a `Host` for the room ID, containing the server name (minus the port) of the
+    /// Returns the host of the room ID, containing the server name (including the port) of the
     /// originating homeserver.
-    ///
-    /// The host can be either a domain name, an IPv4 address, or an IPv6 address.
-    pub fn hostname(&self) -> &Host {
-        &self.hostname
+    pub fn hostname(&self) -> &str {
```
Comment (auscompgeek) [inline src/room_id.rs:47]: It's intended that a future room version will remove server names from room IDs. Do we want to break backwards compat again down the road?

File: src/room_id_or_room_alias_id.rs [hunk 3]
```diff
@@ -32,92 +26,70 @@ use crate::{
 ///     "!n8f893n9:example.com"
 /// );
 /// ```
-#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+#[derive(Clone, Debug, Eq, PartialEq)]
 #[cfg_attr(feature = "diesel", derive(FromSqlRow, QueryId, AsExpression, SqlType))]
 #[cfg_attr(feature = "diesel", sql_type = "Text")]
-pub enum RoomIdOrAliasId {
```
Comment (auscompgeek) [inline src/room_id_or_room_alias_id.rs:38]: Any particular reason for changing this?

File: src/room_id.rs (hunk 2 above)
Comment (jplatte) [reply src/room_id.rs:47]: I wasn't aware of that. Is there an MSC for that?

File: src/room_id_or_room_alias_id.rs (hunk 3 above)
Comment (jplatte) [reply src/room_id_or_room_alias_id.rs:38]: It means I can reuse the `common_impls!` macro, and it removes a bunch of useless branches (for cases where the string representation is used without care for which sigil it starts with).

File: src/room_id.rs (hunk 2 above)
Comment (auscompgeek) [reply src/room_id.rs:47]: I haven't seen any MSC, but I keep hearing about two different possible approaches (using either one-time public keys, or a hash of the m.room.create event). But it very much does sound like it's on the wishlist.

Comment (jplatte) [reply src/room_id.rs:47]: I think I'll keep things the way they are then. It doesn't sound like this is something we'd have to do a semver-breaking release for for a little while, and I'm sure there's going to be other changes in the meantime. Feel free to open an issue about this once there is an MSC though.

**Pull request merged**